Porting account invoice production lot

archetipo opened the pull request:
need review

archetipo committed:
Migration of account_invoice_production_lot

archetipo committed:
[FIX] create decorator

archetipo committed:
[RF]  porting model to API 8.0
[IMP] new test
[ADD] report inovoice, adding SN on invoice report

archetipo committed:
[FIX] test

archetipo committed:
[FIX] html formatting in formatted_note
[FIX] improved test cases
[FIX] Flake8

archetipo committed:
[FIX] contrib.

archetipo committed:
[FIX] test_1_SaleOrder

eLBati commented on account_invoice_production_lot/README.rst:
```diff
@@ -0,0 +1,51 @@
+Usage
+=====
+
+Create a Sale Order, in field  "Create Invoice" set "On Delivery Order".
```
This module should work whatever invoicing policy is set for sale order, doesn't it?
7.0 version did, if I remember correctly

archetipo replied:
you are right! tnx!

archetipo committed:
[FIX] handled all type of   invoicing policy
[IMP] improve tets case

eLBati commented on account_invoice_production_lot/README.rst:
```diff
@@ -0,0 +1,52 @@
+
+Create a Sale Order.
+When the order is accepted and picked out , set production lot
+for each line of products in picking list.
```
> When the order is accepted and picked out , set production lot for each line of products in picking list

I would write some thing like

> When the sale order is accepted and delivery order created, process picking list setting serial numbers on delivered lines

archetipo replied:
ok

eLBati commented on account_invoice_production_lot/invoice.py:
```diff
@@ -0,0 +1,86 @@
+        compute='_get_prod_lots', string="Production Lots")
+
+    displayed_lot_id = fields.Many2one('stock.production.lot', 'Lot')
+    formatted_note = fields.Html('Formatted Note')
```
I would name this field `lot_formatted_note`, to avoid to conflict with other possible generic fields

eLBati commented on account_invoice_production_lot/tests/test_invoice_production_lot.py:
```diff
@@ -0,0 +1,171 @@
+        if split:
+            pick_wizard.item_ids[0].split_quantities()
+            pick_wizard.refresh()
+        for count in range(len(lot_ids)):
```
Why not `for lot_id in lot_ids` ?

eLBati replied:
Oh, I see: it's used in `pick_wizard.item_ids`

eLBati commented on account_invoice_production_lot/tests/test_invoice_production_lot.py:
```diff
@@ -0,0 +1,171 @@
+            self.assertEqual(pick.state, 'assigned')
+            if data:
+                trans = self.run_picking(pick, lot_ids)
+                if trans and pick.action_done():
```
I think the test should fail if this condition is not satisfied

archetipo replied:
I hope so .. if these conditions are not satisfied , the test must fail ..

eLBati replied:
But if line 96 is False, nothing happens and test does not fail

archetipo committed:
[FIX] description in README
[FIX] chenge name of field in lot_formatted_note, to avoid conflict with other possible generic fields
[FIX] test cases if picking action_done fail ,test must fail

archetipo committed:
[FIX] flake 8

pedrobaeza commented on account_invoice_production_lot/README.rst:
```diff
@@ -0,0 +1,52 @@
+Usage
+=====
+
+Create a Sale Order.
```
Use `*` for splitting the points

archetipo replied:
ok thanks!

pedrobaeza commented on account_invoice_production_lot/README.rst:
```diff
@@ -0,0 +1,52 @@
+Invoice Production Lots
+===========
+
+This module prints, for each (customer) invoice line, the delivered production
```
s/prints/shows

pedrobaeza replied:
Why not use it also for incoming lots?

eLBati replied:
@pedrobaeza the purpose of this module is to print invoice report with delivered serial numbers. I can't see the use case where incoming lots could be useful

pedrobaeza replied:
OK, you're right. Forget it.

pedrobaeza commented on account_invoice_production_lot/invoice.py:
```diff
@@ -0,0 +1,86 @@
+from openerp import models, fields, api
+
+
+class account_invoice_line(models.Model):
```
Use CamelCase for the class name

pedrobaeza commented on account_invoice_production_lot/invoice.py:
```diff
@@ -0,0 +1,86 @@
+    def _get_prod_lots(self):
+        if not self.move_line_ids and not self.order_lines:
+            return
+        if self.move_line_ids:
```
You can do `self.mapped('move_line_ids.lot_ids')` for a simpler handling of this

archetipo replied:
Ok thanks a lot

pedrobaeza commented on account_invoice_production_lot/invoice.py:
```diff
@@ -0,0 +1,86 @@
+                if move.lot_ids:
+                    self.prod_lot_ids += move.lot_ids
+        else:
+            for order_line in self.order_lines:
```
The same, use: `self.mapped('order_lines.procurement_ids.move_ids.lot_ids')`

pedrobaeza commented on account_invoice_production_lot/invoice.py:
```diff
@@ -0,0 +1,86 @@
+                        if move.lot_ids:
+                            self.prod_lot_ids += move.lot_ids
+
+    _inherit = "account.invoice.line"
```
Put this just after the class name for better recognition

pedrobaeza commented on account_invoice_production_lot/invoice.py:
```diff
@@ -0,0 +1,86 @@
+        compute='_get_prod_lots', string="Production Lots")
+
+    displayed_lot_id = fields.Many2one('stock.production.lot', 'Lot')
+    lot_formatted_note = fields.Html('Formatted Note')
```
Remove this useless field

eLBati replied:
Actually this is not useless: it contains the HTML to be printed in invoice report. But it can probably become a computed field. This way we can avoid `load_line_lots` and `create` methods. @archetipo what do you think?

pedrobaeza replied:
Ahm, that was the piece I miss for this strange methods/field...

archetipo replied:
@eLBati  @pedrobaeza 
for me is a good idea remove load_line_lots, and make lot_formatted_note a computed field

pedrobaeza commented on account_invoice_production_lot/invoice.py:
```diff
@@ -0,0 +1,86 @@
+        'stock.production.lot',  'stock_prod_lot_invoice_rel', 'invoice_id',
+        compute='_get_prod_lots', string="Production Lots")
+
+    displayed_lot_id = fields.Many2one('stock.production.lot', 'Lot')
```
This is not needed

pedrobaeza commented on account_invoice_production_lot/invoice.py:
```diff
@@ -0,0 +1,86 @@
+        return True
+
+    @api.model
+    def create(self, vals):
```
This code is very confusing and I think not needed (maybe inheritance from the invoice_webkit report module).

pedrobaeza commented on account_invoice_production_lot/invoice.py:
```diff
@@ -0,0 +1,86 @@
+        return res
+
+
+class account_invoice(models.Model):
```
This is also not needed according current logic

pedrobaeza commented on account_invoice_production_lot/invoice.py:
```diff
@@ -0,0 +1,86 @@
+
+    _inherit = "account.invoice.line"
+
+    order_lines = fields.Many2many(
```
Why this isn't a computed field?

eLBati replied:
@pedrobaeza `sale_order_line_invoice_rel` is the table already used by `invoice_lines` field of `sale` module. So, it is filled by `sale` module

pedrobaeza replied:
OK, thanks for the clarification

archetipo committed:
[FIX] remove useless method and field
[FIX] change lot_formatted_note in computed filed
[FIX] test case
[FIX] README

eLBati commented on account_invoice_production_lot/tests/test_invoice_production_lot.py:
```diff
@@ -0,0 +1,172 @@
+        for pick in order.picking_ids:
+            data = pick.force_assign()
+            self.assertEqual(pick.state, 'assigned')
+            if data:
```
If `data` can't be empty, I think you should use `self.assertTrue(data)` and remove this `if`.

eLBati commented on account_invoice_production_lot/tests/test_invoice_production_lot.py:
```diff
@@ -0,0 +1,172 @@
+                trans = self.run_picking(pick, lot_ids)
+                done = pick.action_done()
+                self.assertTrue(done)
+                if trans and done:
```
See above

eLBati commented on account_invoice_production_lot/tests/test_invoice_production_lot.py:
```diff
@@ -0,0 +1,172 @@
+        lot_ids.append(self.getDemoObject('', 'lot_icecream_0'))
+        order = self.getDemoObject('', 'sale_order_0')
+        order.signal_workflow('order_confirm')
+        for pick in order.picking_ids:
```
If `len(order.picking_ids)` must be `1`, I would add `self.assertEqual(len(order.picking_ids), 1)` for clarity

eLBati commented on account_invoice_production_lot/tests/test_invoice_production_lot.py:
```diff
@@ -0,0 +1,172 @@
+        lot_ids.append(self.getDemoObject('', 'lot_icecream_1'))
+        order = self.getDemoObject('', 'sale_order_1')
+        order.signal_workflow('order_confirm')
+        for pick in order.picking_ids:
```
See above

eLBati commented on account_invoice_production_lot/tests/test_invoice_production_lot.py:
```diff
@@ -0,0 +1,172 @@
+        for pick in order.picking_ids:
+            data = pick.force_assign()
+            self.assertEqual(pick.state, 'assigned')
+            if data:
```
See above

eLBati commented on account_invoice_production_lot/tests/test_invoice_production_lot.py:
```diff
@@ -0,0 +1,172 @@
+                trans = self.run_picking(pick, lot_ids, split=True)
+                done = pick.action_done()
+                self.assertTrue(done)
+                if trans and done:
```
See above

eLBati commented on account_invoice_production_lot/tests/test_invoice_production_lot.py:
```diff
@@ -0,0 +1,172 @@
+        lot_ids.append(self.getDemoObject('', 'lot_icecream_0'))
+        order = self.getDemoObject('', 'sale_order_2')
+        order.signal_workflow('order_confirm')
+        for pick in order.picking_ids:
```
See above

eLBati commented on account_invoice_production_lot/tests/test_invoice_production_lot.py:
```diff
@@ -0,0 +1,172 @@
+        for pick in order.picking_ids:
+            data = pick.force_assign()
+            self.assertEqual(pick.state, 'assigned')
+            if data:
```
See above

eLBati commented on account_invoice_production_lot/tests/test_invoice_production_lot.py:
```diff
@@ -0,0 +1,172 @@
+                trans = self.run_picking(pick, lot_ids)
+                done = pick.action_done()
+                self.assertTrue(done)
+                if trans and done:
```
See above

archetipo committed:
[RM] useless button in invoice view
[FIX] test

archetipo commented:
@pedrobaeza  @eLBati  
thanks i've updated the code

pedrobaeza commented:
Two little PEP8 errors:

```
./account_invoice_production_lot/tests/test_invoice_production_lot.py:95:13: F841 local variable 'trans' is assigned to but never used
./account_invoice_production_lot/tests/test_invoice_production_lot.py:125:13: F841 local variable 'trans' is assigned to but never used
```

But the rest :+1:

antoniocanovas commented:
I am testing the module and works fine, but if you print an invoice with no s/n assigned there is a big space between lines (2 lines per page):

![captura de pantalla 2015-08-20 a las 10 27 13](https://cloud.githubusercontent.com/assets/8323017/9379045/c4fec25c-4726-11e5-9f92-2345f221f8f4.png)

archetipo committed:
[FIX] Flake 8
[FIX] Qweb Report

archetipo commented:
@pedrobaeza  thanks!
@konerycloud thanks for notification, i've solved the issue, for testing make sure that Reload from Attachment is False in  Settings /Reports / Invoices

eLBati commented:
:+1: thanks

eLBati commented:
@konerycloud I would merge, if you agree

antoniocanovas commented:
:+1:
